[MAINT] Adapt decoder for scikit-learn > 1.3

ymzayek opened the pull request:
<!---
This is a suggested pull request template for nilearn.
It's designed to capture information we've found to be useful in reviewing pull requests.

If there is other information that would be helpful to include, please don't hesitate to add it!

Please make sure your pull request also follows the
[contribution guidelines](https://nilearn.github.io/stable/development.html#contribution-guidelines) that
will be enforced during the review process.
-->

<!-- Please indicate after the # which issue you're closing with this PR.
This is helpful for the maintainers AND will magically close the issue when this
pull request is merged!
If the PR closes multiple issues, includes "closes" before each one is listed.
https://help.github.com/articles/closing-issues-using-keywords -->
- Closes #4132

<!-- Please give a brief overview of what has changed in the PR.
If you're not sure what to write, consider it a note to the maintainers to indicate
what they should be looking for when they review the pull request. -->
Changes proposed in this pull request:

- Change inheritance of decoder classes
- Move data extraction to `decision_function`
- Fix some docstrings (already not accurate in stable version)
- Check backward compatibility; may need deprecation

ymzayek commented on .github/workflows/numpy2_compatibility.yml:
```diff
@@ -11,7 +11,6 @@ on:
 
 jobs:
     test_numpy2:
-        if: github.repository == 'nilearn/nilearn'
```
TODO: Revert before merging PR

glemaitre commented on nilearn/decoding/decoder.py:
```diff
@@ -1201,7 +1201,7 @@ def __init__(
 
 @fill_doc
-class DecoderRegressor(_BaseDecoder):
+class DecoderRegressor(_BaseDecoder, MultiOutputMixin, RegressorMixin):
```
From the top of the head, the mixins need to be on the most left to not have issue with the MRO (old discussions: https://github.com/scikit-learn/scikit-learn/pull/8022#issuecomment-267091483)

glemaitre commented:
The changes looks good to me apart from the order of inheritance for the MRO.

ymzayek commented:
@glemaitre thanks for pointing that out. I fixed the order though then I had to explicitly define `score` under the two decoder classes. Not sure if there is another way to do it. I added:

```python
    def score(self, X, y, *args):
        """Call score method from base class."""
        return _BaseDecoder.score(self, X, y, *args)
```

Does it look ok now?

bthirion reviewed (approved):
Thx for taking care of that one !

glemaitre commented on nilearn/decoding/decoder.py:
```diff
@@ -1338,6 +1347,10 @@ def __init__(
 
+    def score(self, X, y, *args):
+        """Call score method from base class."""
+        return _BaseDecoder.score(self, X, y, *args)
```
OK I see. In terms of MRO, it means that we favor the `score` define in the mixin instead of the base class. Normally, this is usually what we want. But this is not the case, here. However, if we put the mixin on the right, we will break all the tag system.

Since `_BaseDecoder` is defining the score function, it means that you actually don't want to have the functionality defined by the mixin so you don't need to inherate. So from this point, the right fix to me is the following:

```python
class _BaseDecoder(...):
    def score(self, ...):
        ...
    def _more_tags(self):
        return {"require_y": True}

class Decoder(_BaseDecoder):
    _estimator_type = "classifier"

    ...

class DecoderRegressor(MultiOutputMixin, _BaseDecoder):
    _estimator_type = "regressor"

    ...
```

The use of the `RegressorMixin` and `ClassifierMixin` in your case is just required to defined the `_estimator_type` so let's define it directly in the children class. The second thing that we want is the definition of the `"require_y"` tag that can be done in the base class since this is common to both mixins.

I did not foresee the part or redefining the `score` function. Let me know if this is making sense to you @ymzayek.

ymzayek replied:
> I did not foresee the part or redefining the score function. Let me know if this is making sense to you @ymzayek.

I did not either and then I tried to go around it with redefining `score` but you are right, we do not need those mixins and can just define the needed attribute and tag function as you suggest. I was coming to the same conclusion so makes total sense. Thanks!

I'll make the changes and then I think it should be good to go

glemaitre commented on nilearn/decoding/decoder.py:
```diff
@@ -1065,6 +1065,9 @@ def _predict_dummy(self, n_samples):
         return scores.ravel() if scores.shape[1] == 1 else scores
 
+    def _more_tags(self):
+        return {"require_y": True}
```
You can easily cover this line by adding a check that `require_y` is set to True and `mulitoutput` is False for the classifier and True for the regressor. It allows to check that the MRO works as expected as well.

Remi-Gau reviewed (approved):
Looks good to me

ymzayek commented on nilearn/decoding/decoder.py:
```diff
@@ -870,6 +871,9 @@ def decision_function(self, X):
         y_pred: ndarray, shape (n_samples,)
             Predicted class label per sample.
         """
+        # for backwards compatibility
+        if not isinstance(X, type(np.array)):
+            X = self.masker_.transform(X)
```
Sorry I though it was ready but then realized that this change to take X as a niimg was not backward compatible so I added this conditional. However, the docstring says X is a niimg like object so it can also be seen as a bug. Should we deprecate behavior here instead? Not sure how often people call this function directly but it is not even tested directly in our test suite

glemaitre commented on nilearn/decoding/decoder.py:
```diff
@@ -870,6 +871,9 @@ def decision_function(self, X):
             Predicted class label per sample.
         """
+        # for backwards compatibility
+        if not isinstance(X, type(np.array)):
```
This is looking weird to me. Do you mean:

```python
if not isinstance(X, np.ndarray):
    ...
```

Usually, having `type` in an `isinstance` looks weird (sometimes it is because you intend to call `issubclass`).

ymzayek replied:
Yes you are right, thanks!

ymzayek commented:
If there are no more comments I will merge this when the CI finishes
